test/integration/filters/decode_headers_return_stop_all_filter.cc

```diff
@@ -0,0 +1,122 @@
+    Http::HeaderEntry* entry_added = header_map.get(Envoy::Http::LowerCaseString("added_size"));
+    ASSERT(entry_content != nullptr && entry_added != nullptr);
+    content_size_ = std::stoul(std::string(entry_content->value().getStringView()));
+    added_size_ = std::stoul(std::string(entry_added->value().getStringView()));
+    Http::HeaderEntry* entry_is_first_trigger =
+        header_map.get(Envoy::Http::LowerCaseString("first_trigger"));
+    is_first_trigger_ = entry_is_first_trigger != nullptr;
+    // Remove "first_trigger" headers so that if the filter is registered twice in a filter chain,
+    // it would act differently.
+    header_map.remove(Http::LowerCaseString("first_trigger"));
+
+    createTimerForContinue();
+
+    Http::HeaderEntry* entry_buffer = header_map.get(Envoy::Http::LowerCaseString("buffer_limit"));
+    if (entry_buffer == nullptr || !is_first_trigger_) {
+      return Http::FilterHeadersStatus::StopAllIterationAndBuffer;
+    } else {
+      watermark_enabled_ = true;
+      buffer_limit_ = std::stoul(std::string(entry_buffer->value().getStringView()));
+      decoder_callbacks_->setDecoderBufferLimit(buffer_limit_);
+      header_map.remove(Http::LowerCaseString("buffer_limit"));
+      return Http::FilterHeadersStatus::StopAllIterationAndWatermark;
+    }
+  }
+
+  Http::FilterDataStatus decodeData(Buffer::Instance& data, bool) override {
+    ASSERT(timer_triggered_);
+    if (is_first_trigger_) {
+      if (watermark_enabled_) {
+        // High watermark reached before all data are received. The rest of the data is sent after
+        // iteration resumes.
+        EXPECT_LT(data.length(), content_size_);
+      } else {
+        // decodeData will only be called once after iteration resumes.
+        EXPECT_EQ(data.length(), content_size_);
+      }
+      Buffer::OwnedImpl added_data(std::string(added_size_, 'a'));
+      decoder_callbacks_->addDecodedData(added_data, false);
+    } else {
+      EXPECT_TRUE(data.length() == content_size_ + added_size_ ||
+                  data.length() == content_size_ + added_size_ * 2);
+    }
+    return Http::FilterDataStatus::Continue;
+  }
+
+  Http::FilterTrailersStatus decodeTrailers(Http::HeaderMap&) override {
+    ASSERT(timer_triggered_);
+    if (is_first_trigger_) {
+      Buffer::OwnedImpl data(std::string(added_size_, 'a'));
+      decoder_callbacks_->addDecodedData(data, false);
+    }
+    return Http::FilterTrailersStatus::Continue;
+  }
+
+private:
+  // Creates a timer to continue iteration after conditions meet.
+  void createTimerForContinue() {
+    delay_timer_ = decoder_callbacks_->dispatcher().createTimer([this]() -> void {
+      // If decodeHeaders() returns StopAllIterationAndBuffer, triggers the timer when all the
+      // request data has been received. If decodeHeaders() returns StopAllIterationAndWatermark,
+      // triggers the timer when received data exceed buffer limit.
+      if ((content_size_ > 0 &&
```

Comment:
I think buffer_limit should not be 0 otherwise that says we don't want to buffer any data. content_size can be 0. It just means we send 0 data, :)

Reply:
This filter is designed to call continueDecoding() when all the data have been received. It is for the ease of test of StopAll status, not for checking if continueDecoding() is called. I think it should be OK, since it's testing only, :)